algorithms/src/snark/marlin/ahp/prover/message.rs

```diff
@@ -14,18 +14,26 @@
 // You should have received a copy of the GNU General Public License
 // along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
 
-use snarkvm_fields::Field;
+use snarkvm_fields::PrimeField;
 use snarkvm_utilities::{error, serialize::*, ToBytes, Write};
 
-/// The prover message in the third round.
+use crate::snark::marlin::CircuitId;
+use std::collections::BTreeMap;
+
 #[derive(Clone, Debug, Default, PartialEq, Eq, CanonicalSerialize, CanonicalDeserialize)]
-pub struct ThirdMessage<F: Field> {
+pub struct MatrixSums<F: PrimeField> {
     pub sum_a: F,
     pub sum_b: F,
     pub sum_c: F,
 }
 
-impl<F: Field> ToBytes for ThirdMessage<F> {
+/// The prover message in the third round.
+#[derive(Clone, Debug, Default, PartialEq, Eq, CanonicalSerialize, CanonicalDeserialize)]
+pub struct ThirdMessage<F: PrimeField> {
+    pub sums: BTreeMap<CircuitId, MatrixSums<F>>,
+}
```

Comment:
We should be careful about how this is serialized in the final proof (we don't want to serialize the `CircuitId` anywhere.